components/camel-milo/src/main/java/org/apache/camel/component/milo/client/internal/SubscriptionManager.java

```diff
@@ -368,6 +379,197 @@ public void dispose() {
+                            failedFuture.completeExceptionally(new IllegalArgumentException(
+                                    format(
+                                            "Invalid number of browse results: %s, expected %s", partials.size(),
+                                            browseDescriptions.size())));
+                            return failedFuture;
+                        }
+
+                        final List<CompletableFuture<Map<ExpandedNodeId, BrowseResult>>> futures = new ArrayList<>();
+
+                        for (int i = 0; i < partials.size(); i++) {
+
+                            futures.add(browse(browseDescriptions.get(i), partials.get(i), depth, maxDepth, pattern,
+                                    maxNodesPerRequest));
+                        }
+
+                        return flatten(futures);
+                    });
+        }
+
+        // Wrapper for looking up nodes and instantiating initial browse descriptions according to the configuration provided
+        @SuppressWarnings("unchecked")
+        public CompletableFuture<Map<ExpandedNodeId, BrowseResult>> browse(
+                List<ExpandedNodeId> expandedNodeIds, BrowseDirection direction, int nodeClasses, int depth, int maxDepth,
+                Pattern pattern, boolean includeSubTypes, int maxNodesPerRequest) {
+
+            final CompletableFuture<NodeId>[] futures = expandedNodeIds.stream().map(this::lookupNamespace)
+                    .toArray(CompletableFuture[]::new);
+
+            return CompletableFuture.allOf(futures)
+
+                    .thenCompose(__ -> {
+
+                        List<NodeId> nodeIds = Stream.of(futures).map(CompletableFuture::join)
+                                .collect(Collectors.toList());
+
+                        return completedFuture(nodeIds.stream().map(nodeId -> new BrowseDescription(
+                                nodeId, direction, Identifiers.References, includeSubTypes, uint(nodeClasses),
+                                uint(BrowseResultMask.All.getValue()))).collect(Collectors.toList()));
+                    })
+
+                    .thenCompose(descriptions -> browse(descriptions, depth, maxDepth, pattern, maxNodesPerRequest))
+
+                    .whenComplete((actual, error) -> {
+
+                        if (!LOG.isErrorEnabled()) {
+
+                            return;
+                        }
+
+                        final String expandedNodeIdsString = expandedNodeIds.stream()
+                                .map(ExpandedNodeId::toParseableString)
+                                .collect(Collectors.joining(", "));
+
+                        if (actual != null) {
+                            LOG.debug("Browse node(s) {} -> {} result(s)", expandedNodeIdsString, actual.size());
+
+                        } else {
+                            LOG.error("Browse node(s) {} -> failed: {}", expandedNodeIdsString, error);
```

Comment:
Can we do something else than just log an error? Set an exception on the exchange as a result that this failed

Reply:
Yes, makes sense, of course. I changed that - moved it to the producer, that reduces the complexity in the SubscriptionManager class, too.